---
title: "03-Integrative_analysis"
output: html_notebook
---

# Description

In this notebook an integrated analysis is preformed.
Pair interaction data of miR - target(mRNA) and DNAm - target(mRNA) is an output from notebook 02-Regulation_targets.
Integrative analysis is goint go be focused on mRNA so a table with significantly DE and one of its regulator. Since there are some genes that have multiple regulators mir’s with the best correlation will be selected as a regulator. For DNAm !!! get back to this
To explore the influence of miR and DNAm on mRNA expression we will preform GSEA on groups of genes that are under regulation of DNAm, mir or both. Next, we will make an unsupervised clustering on genes with FC features to find patterns in gene expressions. (OVO ISTO POBOLJŠATI)

# Libraries
```{r}
library(venn)
library(dplyr)
library(GenomicRanges)
library(gprofiler2)
library(DESeq2)
library(tibble)
library(kohonen)
library(rprojroot)



# source(file = file.path(getwd(), "scripts", "idMap.R"))
# source(file = file.path(getwd(), "scripts", "integrateOmicsData.R"))
# source(file = file.path(getwd(), "scripts", "kmeansHeatmap.R"))

genes_dir <- file.path(find_rstudio_root_file(), "Scripts", "genes.hg19.ensembl.rds")
```


# Loading data
```{r Libraries, message=FALSE}
dir_dest1 <- "/home/katarina/MyProjects/HNSCC/01-Exploratory_analysis_and_differential_expression"
dir_dest2 <- "/home/katarina/MyProjects/HNSCC/02-Regulation_targets"

## 01
# mRNA
mRNA_res <- readRDS(file.path(dir_dest1,"outputs/RDS/mRNA_res.rds"))
# miR
mir_res <- readRDS(file.path(dir_dest1,"outputs/RDS/mir_res.rds"))
# dnam
dnam_res_cgi <- readRDS(file.path(dir_dest1,"outputs/RDS/dnam_res_cgi.rds"))


## 02
# miR-mRNA
targ_mir_mRNA <- readRDS(file.path(dir_dest2,"outputs/RDS/targ_mir_mRNA.rds"))
cor_mir <- readRDS(file.path(dir_dest2,"outputs/RDS/cor_mir.rds"))
# dnam-mRNA
targ_dnam_mRNA <- readRDS(file.path(dir_dest2,"outputs/RDS/targ_dnam_mRNA.rds"))
```


# Integrated table
In this chapter we build the integrated table that will contain multiple information from different data sets.

Firstly, filtering mRNA results to get only significantly DE genes.
```{r}
mRNA_de <- filter(as.data.frame(mRNA_res),padj < 0.05)
```

# Venn
```{r}
simple_tb <- 
  data.frame(genes=rownames(mRNA_de)) %>%
  mutate(mir=dplyr::case_when(
    is.element(genes,targ_mir_mRNA$target_ensembl) ~ 1, TRUE ~ 0)) %>%
  mutate(dnam=dplyr::case_when(
    is.element(genes, targ_dnam_mRNA$ensembl) ~ 1, TRUE ~ 0))

# table of genes under regulation
table(simple_tb [,c("mir","dnam")])



venn::venn(x =list(simple_tb[simple_tb$dnam ==1,"genes"],
             simple_tb[simple_tb$mir ==1,"genes"]), 
     zcolor = "style",
     snames = c("DNA methylated","microRNA"))
```
## GSEA analysis on simple_tb
To explore the influence of different types of regulation we preform gsea on crude groups of genes where the genes are presneted in the venn.
From the gostplot there are not many terms associated with genes that are ONLY under DNA methylation regulation (Only muscle contraction -REAC). In the intersection part there are terms that are involved in cell cycle regulation, regulation of transcription, .. Genes under mir regulation have many terms involved with the immune system, also some terms involved with cell mobility (cell-cell adhesion, cell migration, collagen degradation). The rest that are not under dnam or mir regulation have terms involved in muscle contractions and some immunological pathways. TF database added later, comment: In all groups there are many hits with TF.

```{r}
# Building multiple queries
query_ls <- 
  list(dnam = filter(simple_tb, mir == 0 & dnam == 1 )$genes,
       mir = filter(simple_tb, mir == 1 & dnam == 0 )$genes,
       intersection = filter(simple_tb, mir == 1 & dnam == 1 )$genes,
       rest = filter(simple_tb, mir == 0 & dnam == 0 )$genes)

# Sources
GSEA_SOURCES <- c("GO:BP","CORUM","KEGG","REAC","WP","MIRNA", "TF")

gost <-
  gost(query = query_ls,
       organism= "hsapiens",
       exclude_iea=TRUE,
       domain_scope="annotated",
       ordered_query=F,
       sources=GSEA_SOURCES)

# Plot and table of significant results
gostplot(gost, capped = T, interactive = T)
```

# Regulation table - integrated
Building a complete and comprehensive table with results of DE and DM of the transcriptomic (mRNA), small-RNA transcriptomic (mir) and DNA methylation (CpG islands) to have all significant data in one data frame. The comprehensive table will contain all mir mRNA and dnam mRNA pairs. ! add colnames!

```{r}
# this steps for this are described in the script
integ_df <-
  integrateOmicsData(mrna_res_obj = mRNA_res,
                   mir_res_obj = mir_res,
                   dnam_res_obj = dnam_res_cgi,
                   mir_targets_cor = cor_mir,
                   genes_file = genes_dir)

head(integ_df)
message(paste0("Dimensions of integrated tbl: ", dim(integ_df)[1]))
# Summary statistics
summary(integ_df)
```

## Subseting integrated table
Here we want to make a table that has only one row per mRNA so that we can explore the influence of DE mir and dnam that are putative regulators of said mRNA. For multiple mir regulators we choose the best correlating one. For multiple dm cgi’s involved in regulating an mRNA we choose the one with the “best” (larger absolute value) fold change.
```{r}
reg_df <-
  integ_df %>%
  # Group by gene to compare their regulators
  group_by(gene) %>%
  # Storing the best correlating mir into best_cor column
  # NA is for those that are not under mir regulation
  mutate(best_cor=ifelse(!is.na(mir_r), mir[which.max(abs(mir_r))], NA)) %>%
  # filtering out mirs that are not the best putative regulators
  dplyr::filter(is.na(mir) | mir == best_cor)  %>%
  # Storing the cgi with max abs value in best_cgi
  mutate(best_cgi= ifelse(!is.na(dnam_quot),                                 dnam_condensed_ranges[which.max(abs(dnam_quot))],
                                 NA)) %>%
  # Filtering out cgi that are not the best
  dplyr::filter(is.na(dnam_condensed_ranges) |
                  dnam_condensed_ranges == best_cgi) %>%
  # Dropping uneeded columns. This info is already in mir and dnam ranges
  dplyr::select(-best_cor, -best_cgi) %>%
  ungroup()

reg_df
```
# Clustering
Using two unsupervised clustering methods SOM and kmeans to explore clusters of genes under different combination of regulations. Features are fold changes of mRNA and mir data and quotient of beta intensity value for dna methylation. We hypothesized that groups with mRNA FC values will be opposite of mir and dnam regulators which would suggest that those groups are under regulation of mir and/or dnam.

## **kmeans**
K means clusters mRNA genes into k number of clusters in which each mRNA belongs to the cluster with the nearest mean (or cluster centroid). For more details look at the lecture materials from pmf machine learning.
```{r}
# Prepping kmeans table
k_tb <- data.frame(row.names = reg_df$gene,
                   dplyr::select(reg_df,
                                 gene_FC,
                                 mir_FC,
                                 dnam_quot))

hist(k_tb$dnam_quot)
# Scaling (standardization - x-mean/sd)
# k_tb <- scale(k_tb) %>%
#   as.data.frame()


k_tb$dnam_quot <- scale(k_tb$dnam_quot)

hist(k_tb$dnam_quot)

# All NA values are 0
k_tb[is.na(k_tb)] <- 0
```


```{r}
kmeansHeatmap(k_tb, clusters=8:12, nstart =50)
```



# Session Info
```{r Session info}
sessionInfo()
```
