---
title: "Plots for paper"
author: "Katarina MandiÄ‡"
date: "`r format(Sys.time(), '%d %b, %Y')`"
output:
  html_document:
    css: style.css
---
 

# Description
Plotting for the paper

# Libraries
```{r Libraries, message=FALSE}
library(ggplot2)
library(DESeq2)
library("pheatmap")
library("RColorBrewer")
library(dplyr)
library(gprofiler2)
library(plyranges)
library(rprojroot)
library(RnBeads)
library(tidyr)

# Leads to the project directory
project_dir <- find_rstudio_root_file()

source(file = file.path(project_dir,"src", "utils.R"))

genes_dir <- file.path(project_dir, "data","input", "genes.hg19.ensembl.rds")
```

Helpers
```{r}
calculate_shared_percentage <- function(vector1, vector2) {
  if (length(vector1) == 0 || length(vector2) == 0) {
    return(0)  # If either vector is empty, the percentage is 0.
  }
  
  # Convert vectors to sets for efficient intersection calculation.
  set1 <- unique(vector1)
  set2 <- unique(vector2)
  
  # Calculate the intersection of the two sets.
  intersection <- intersect(set1, set2)
  
  # Calculate the percentage of shared elements.
  percentage <- length(intersection) / length(union(set1, set2)) * 100
  
  return(round(percentage,2))
}
```

# Load data
```{r}
mRNA_dds <- readRDS(file = file.path(project_dir, "data/processed/mRNA_dds.rds"))
mRNA_res <- readRDS(file.path(project_dir, "data/processed", "mRNA_res.rds"))
                      
mir_dds <- readRDS(file = file.path(project_dir, "data/processed/mir_dds.rds"))
rnb_set <- readRDS(file = file.path(project_dir, "data/processed/rnb_set.rds"))

# This table is the one for the paper, but contains all samples so its usefull
samples_ppr <- readxl::read_xlsx(file.path(project_dir, "data/input/samples.xlsx"))

# Full annotation
annot_full <- readxl::read_xlsx(file.path(project_dir, "data/raw/hnscc_annot.xlsx"))
```


# PCA plot on mRNA data
```{r}
mRNA_rld <- DESeq2::rlog(mRNA_dds)
```


```{r}
# selecting data for plotting pca
mRNA_pcadata <- 
  DESeq2::plotPCA(mRNA_rld, 
                  intgroup = c("Hist", "HPV", "Tissue"),
                  returnData = TRUE,
                  ntop = 18104) %>%
  dplyr::rename(Sample ="Hist") %>%
  mutate(HPV = ifelse(HPV==1, "P","N"))
# selecting variance
percentVar <- round(100 * attr(mRNA_pcadata, "percentVar"))


# Color for PCA 
col_pca <- c(
  # Dark red
  brewer.pal(8, "Reds")[7],
  brewer.pal(8, "Blues")[7]
)


# plotting pca data
ggplot(mRNA_pcadata, aes(x = PC1, y = PC2, color = Tissue, shape = HPV, fill = Sample)) +
  geom_point(size =4,
             stroke =1.2) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
   theme(#panel.grid = element_blank(), 
         # panel.background = element_rect(fill = brewer.pal(3,"Greys")[1]),
     panel.background = element_rect(fill = "white"),
         panel.border = element_rect(fill = "transparent"),
         panel.grid = element_line(color = brewer.pal(5,"Greys")[1]),
         legend.key=element_blank(),
         axis.title=element_text(size=18),
         legend.text=element_text(size=18),
         legend.title=element_text(size=18))+
  coord_fixed() +
  #ggtitle("PCA with rlog data")+
  scale_shape_manual(values = c(21, 22) )+
  scale_color_manual(values = c("black", brewer.pal(5,"Greys")[3]) )+
  scale_fill_manual(values= col_pca)+
  guides(fill=guide_legend(override.aes=list(shape=21)))+
  guides(color=guide_legend(override.aes=list(shape=21))) +
  guides(fill=guide_legend(override.aes=list(shape=21,fill=c(col_pca))))

  #geom_text_repel(aes(label = colData(mRNA_rld)$Sample.ID), size =3)
  #geom_text_repel(aes(label = colData(mRNA_rld)$GRCE.NAME), size =3)

```

Saving to file
```{r, eval=FALSE}
ggsave(height=10, width = 10, filename = file.path(project_dir, "data/output/figures/PCA_mRNA.pdf"))
```

### Higher PC 
```{r}
# Assume df is your data frame
pre_pca_data <-  assay(mRNA_rld)

# Calculate variance of each row
variances <- apply(pre_pca_data, 1, var)

# Sort rows by variance
pre_pca_data <- pre_pca_data[order(variances, decreasing = TRUE), ]

# Select top 500 variable rows
pre_pca_data <- pre_pca_data[1:500, ]

# Uncomment this to do pca on whole gene expression data
# pre_pca_data <-  assay(mRNA_rld)
pca_res <- prcomp(t(pre_pca_data))

mRNA_pcadata <- as.data.frame(pca_res$x)

# Color for PCA 
col_pca <- c(
  # Dark red
  brewer.pal(8, "Reds")[7],
  brewer.pal(8, "Blues")[7]
)

col_pca_2 <- c(
  brewer.pal(8, "Blues")[7], 
  brewer.pal(8, "Reds")[2],
  brewer.pal(8, "Reds")[7]
)

mRNA_pcadata$Sample <- as.vector(colData(mRNA_rld)$Hist)
mRNA_pcadata$Tissue <- colData(mRNA_rld)$Tissue
mRNA_pcadata$HPV <- ifelse(colData(mRNA_rld)$HPV == 1, "P","N")
samples <- rownames(colData(mRNA_rld)) 
mRNA_pcadata$HPV_ctat <- ifelse(colData(mRNA_rld)$HPV == 0, "N", ifelse(is.element(samples, c("KBD3", "KBD15", "KBD17")), "PP", "P"))

# Get the variance explained by each principal component
percentVar <- round(100 * (pca_res$sdev^2 / sum(pca_res$sdev^2)), 1)
```


```{r}
plot_list <- list()

# Loop over the first 10 principal components
for(i in seq(1, 12, by = 2)) {

    j = i+1
    # plotting pca data
    p <-
    # ggplot(mRNA_pcadata, aes_string(x = paste0("PC", i), y = paste0("PC", j), color = "Tissue", shape = "HPV", fill = "Sample")) +
    # For extra annotation
    ggplot(mRNA_pcadata, aes_string(x = paste0("PC", i), y = paste0("PC", j), color = "Tissue", shape = "Sample", fill = "HPV_ctat")) +
      geom_point(size =4, stroke =1.2) +
      xlab(paste0("PC", i, ": ", percentVar[i], "% variance")) +
      ylab(paste0("PC", j, ": ", percentVar[j], "% variance")) +
      theme(panel.background = element_rect(fill = "white"),
            panel.border = element_rect(fill = "transparent"),
            panel.grid = element_line(color = brewer.pal(5,"Greys")[1]),
            legend.key=element_blank(),
            axis.title=element_text(size=18),
            legend.text=element_text(size=18),
            legend.title=element_text(size=18),
            legend.position = "none") +  # Add this line to remove the legend
      coord_fixed() +
      scale_shape_manual(values = c(21, 22) )+
      scale_color_manual(values = c("black", brewer.pal(5,"Greys")[3]) )+
      scale_fill_manual(values= col_pca_2)+
      guides(fill=guide_legend(override.aes=list(shape=21)))+
      guides(color=guide_legend(override.aes=list(shape=21))) +
      guides(fill=guide_legend(override.aes=list(shape=21, fill=c(col_pca))))
    
    
    # Add the plot to the list
    plot_list[[length(plot_list) + 1]] <- p
    
    # ggsave(filename = file.path(project_dir, "data/output/figures/PCA_plots", paste0("PC",i,"_","PC",j, ".png")))
    # 
}

grid_plot <- do.call(arrangeGrob, c(plot_list, ncol = 3))

ggsave(width = 10, height = 7,grid_plot, filename = file.path(project_dir, "data/output/figures/PCA_plots", "higher_PCA_grid_all_horizontal_ctat_annot.png"))

```


### Distnace
Sample distance matrix
```{r}

sampleDist_mRNA <- stats::dist(t(pre_pca_data), method = "euclidean")
# sampleDist_mRNA <- stats::dist(t(assay(mRNA_rld)), method = "manhattan")
sampleDist_mRNA <- as.data.frame(as.matrix(sampleDist_mRNA))
```


```{r}
annot_data <- data.frame(Sample = as.vector(colData(mRNA_rld)$Hist)) 
rownames(annot_data) <- rownames(sampleDist_mRNA)
annot_data$Tissue <- colData(mRNA_rld)$Tissue
annot_data$HPV <- ifelse(colData(mRNA_rld)$HPV == 1, "P","N")

colors <- rev(brewer.pal(8,"Blues"))

annot_col <- list(
  # Tumor colors
  Sample = setNames(c(brewer.pal(8, "Reds")[7],  brewer.pal(8, "Blues")[7]),
                   c("cancer", "control")),
  # HPV colors
  HPV = setNames( brewer.pal(8, "Paired")[c(3,4)], 
                  c("P","N")),
  # Tissue colors
  Tissue = setNames(brewer.pal(8, "Paired")[c(7,8)],
                    c("OP","O")),
    # Smoking drinking colors
  Smoking_Drinking = setNames( c("forestgreen","gold", "white"), 
                  c("0","1", "?")),
  Age = brewer.pal(9, "Greys"),
  Gradus = setNames(brewer.pal(9, "Reds")[c(1,3,5,7,9)],
                    c("nema nalaza", "0", "1", "2", "3"))
)

pheatmap(sampleDist_mRNA,
         # clustering_distance_rows = sampleDist_mir,
         # clustering_distance_cols = sampleDist_mir,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         col = colors,
         # treeheight_col = 0,
         treeheight_row = 0,
         # annotation_row = annot_data,
         annotation_col = annot_data,
         annotation_colors = annot_col,
         show_colnames = FALSE,
         show_rownames = TRUE,
         annotation_names_col = TRUE,
         filename = file.path(project_dir, "data/output/figures/mRNA_sample_dist_euclid_dend.png"))
```


# miR distance
Since I made a presentable PCA plot for mRNA data I will first make sample distance plot for miR
```{r }
# Variance stabilizng 
mir_rld <- rlog(mir_dds)
# calculating sample distance matrix
sampleDist_mir <- stats::dist(t(assay(mir_rld)), method = "euclidean")
sampleDist_mir <- as.data.frame(as.matrix(sampleDist_mir))
```



```{r}
colors <- colorRampPalette( rev(brewer.pal(9, "Oranges")) )(255)

annot_col <- list(
  # Tumor colors
  Hist = setNames(c(brewer.pal(8, "Reds")[7],  brewer.pal(8, "Blues")[7]),
                   c("cancer", "control")),
  # HPV colors
  HPV = setNames( brewer.pal(8, "Paired")[c(3,4)], 
                  c("P","N")),
  # Tissue colors
  Tissue = setNames(brewer.pal(8, "Paired")[c(7,8)],
                    c("OP","O")),
    # Smoking drinking colors
  Smoking_Drinking = setNames( c("forestgreen","gold", "white"), 
                  c("0","1", "?")),
  Age = brewer.pal(9, "Greys"),
  Gradus = setNames(brewer.pal(9, "Reds")[c(1,3,5,7,9)],
                    c("nema nalaza", "0", "1", "2", "3"))
)

annot_data <- colData(mir_rld) %>% 
  as.data.frame() %>%
  dplyr::select(Hist, Tissue, HPV) %>%
  # changing normal to control in Hist
  mutate(Hist = ifelse(Hist == "normal", "control", "cancer"))

# additional annotations
# annot_data <-
#   left_join(tibble::rownames_to_column(annot_data), 
#             # first group of annot
#           dplyr::select(annot_full, Sample_ID, Smoking_Drinking, Age, Gradus),
#           # second group
#           # select(annot_full, Sample_ID, Stage, Smoking_Drinking, Gradus),
#           by = c("rowname" = "Sample_ID")) %>%
#   tibble::column_to_rownames("rowname")


pheatmap(sampleDist_mir,
         # clustering_distance_rows = sampleDist_mir,
         # clustering_distance_cols = sampleDist_mir,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         col = colors,
         treeheight_col = 0,
         treeheight_row = 0,
         #annotation_row = annot_data,
         annotation_col = annot_data,
         annotation_colors = annot_col,
         show_colnames = FALSE,
         show_rownames = FALSE,
         annotation_names_col = FALSE)
```

Saving to file
```{r, eval=FALSE}
pheatmap(sampleDist_mir,
         # clustering_distance_rows = sampleDist_mir,
         # clustering_distance_cols = sampleDist_mir,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         col = colors,
         treeheight_col = 0,
         treeheight_row = 0,
         #annotation_row = annot_data,
         annotation_col = annot_data,
         annotation_colors = annot_col,
         show_colnames = FALSE,
         show_rownames = FALSE,
         annotation_names_col = FALSE,
         fontsize = 30,
         width = 15,
         height = 15,
         cellwidth = 40,
         cellheight = 40,
         # filename = file.path(project_dir, "data/output/figures/extra_annot_sampleDistance_miR.png"))
         filename = file.path(project_dir, "data/output/figures/mir_distance_heatmap.pdf"))
```


# DNAm distance
```{r}
# Novariance stabilizarion?
# calculating sample distance matrix
sampleDist_dnam <- stats::dist(t(meth(rnb_set)))
sampleDist_dnam <- as.data.frame(as.matrix(sampleDist_dnam))
```

```{r}
colors <- colorRampPalette( rev(brewer.pal(9, "Greens")) )(255)
# I have a full list in mir
# annot_col <- list(
#   # Tumor colors
#   Hist = setNames(c(brewer.pal(8, "Reds")[7],  brewer.pal(8, "Blues")[7]),
#                    c("cancer", "control")),
#   # HPV colors
#   HPV = setNames( c("forestgreen","gold"), 
#                   c("P","N")),
#   # Tissue colors
#   Tissue = setNames(c(brewer.pal(8, "Dark2")[3], brewer.pal(8, "Dark2")[7]),
#                     c("OP","O")),
#   # Smoking drinking colors
#   Smoking_Drinking = setNames( c("forestgreen","gold", "pink"), 
#                   c("0","1", "?"))
# )



annot_data <- pheno(rnb_set) %>% 
  as.data.frame() %>%
  dplyr::mutate(Hist = ifelse(Sample_group == "Cancer", "cancer", "control")) %>%
  dplyr::select(Sample_ID, Hist, HPV) %>%
  # Adding tissue locations from sample_ppr table 
  left_join(dplyr::select(samples_ppr,ID, Tissue = "Location") %>%
              mutate(ID=tolower(ID)),
            by = c("Sample_ID" = "ID")) %>%
  dplyr::select(Hist, Tissue, HPV)

rownames(annot_data) <- pheno(rnb_set)$Sample_ID



# additional annotations
# annot_data <-
#   left_join(tibble::rownames_to_column(annot_data), 
#             # first group of annot
#           dplyr::select(annot_full, Sample_ID, Smoking_Drinking, Age, Gradus) %>%
#           # second group
#           # select(annot_full, Sample_ID, Stage, Smoking_Drinking, Gradus) %>%
#             mutate(Sample_ID = tolower(Sample_ID)),
#           by = c("rowname" = "Sample_ID")) %>%
#   tibble::column_to_rownames("rowname")

pheatmap(sampleDist_dnam,
         # clustering_distance_rows = sampleDist_dnam,
         # clustering_distance_cols = sampleDist_dnam,
         col = colors,
         treeheight_col = 0,
         treeheight_row = 0,
         #annotation_row = annot_data,
         annotation_col = annot_data,
         annotation_colors = annot_col,
         show_colnames = FALSE,
         show_rownames = FALSE,
         annotation_names_col = FALSE)

```

Saving to file
```{r, eval=FALSE}
pheatmap(sampleDist_dnam,
         # clustering_distance_rows = sampleDist_dnam,
         # clustering_distance_cols = sampleDist_dnam,
         col = colors,
         treeheight_col = 0,
         treeheight_row = 0,
         #annotation_row = annot_data,
         annotation_col = annot_data,
         annotation_colors = annot_col,
         show_colnames = FALSE,
         show_rownames = FALSE,
         annotation_names_col = FALSE,
         fontsize = 30,
         width = 16,
         height = 15,
         cellwidth = 40,
         cellheight = 40,
         # filename =file.path(project_dir, "data/output/figures/extra_annot_sampleDistance_dnam.png"))
         filename = file.path(project_dir, "data/output/figures/dnam_distance_heatmap.pdf"))
```


# DE expression gsea

## Loading data
```{r}
mRNA_res <- readRDS(file.path(project_dir, "data/processed/mRNA_res.rds"))

```


```{r}
mRNA_gost <- as.data.frame(mRNA_res)%>%
  filter(padj < 0.05) %>%
  rownames() %>%
  unique()


# Using only part of sources
GSEA_SOURCES <- c("GO:BP","KEGG","REAC")


# Call on the gost 
mRNA_gost <- 
  gost(mRNA_gost,
       organism= "hsapiens",
       exclude_iea=TRUE,
       domain_scope="annotated",
       ordered_query=F,
       sources=GSEA_SOURCES)

# Change query name
mRNA_gost$result$query <- ""

highlighted_gsea_terms <- read.csv(file.path(project_dir,"data/input/GSEA_terms.txt"),
                                   header = F)[,1]

#Plot and table of significant results
gostplot(mRNA_gost ,capped = F, interactive = T)

publish_gostplot(gostplot(mRNA_gost ,capped = F, interactive = F),
                 highlight_terms = highlighted_gsea_terms)

```

```{r, eval=FALSE}
publish_gostplot(gostplot(mRNA_gost ,capped = F, interactive = F),
                 highlight_terms = highlighted_gsea_terms,
                 filename = file.path(project_dir, "data/output/figures/gost_plot_mRNA_highlighted.png"))
```


# mir targets histograms
will not be doing that
## Loading data
```{r, eval=FALSE}
targ_mir_mRNA <- 
  readRDS( "../02-Regulation_targets/outputs/RDS/targ_mir_mRNA.rds")

cor_mir <-
  readRDS("../02-Regulation_targets/outputs/RDS/cor_mir.rds")
```

## plot
```{r, eval=FALSE}
mir_freq <- 
  table(cor_mir$mir) %>%
  as.data.frame()


ggplot(mir_freq, aes(Freq))+
  geom_histogram()+
  stat_bin(binwidth = 1, bins = 50)
```
# Mir mRNA scatterplot




# DM CpGs
## Loading data
```{r}
dnam_res_sites <- readRDS(file.path(project_dir,"data/processed/dnam_res_sites.rds")) %>%
  as.data.frame()

rnb_set <- readRDS(file = file.path(project_dir, "data/processed/rnb_set.rds"))
```


```{r}
dnam_res_sites_annotated <- 
  annotation(rnb_set)[(dnam_res_sites$diffmeth.p.adj.fdr < 0.05),] %>%
  dplyr::select(Chromosome, Start, End) %>%
  as_granges(seqnames = Chromosome, start = Start, end = End, keep_mcols = TRUE)

mcols(dnam_res_sites_annotated)$mean.quot.log2 <-
  dnam_res_sites[(dnam_res_sites$diffmeth.p.adj.fdr< 0.05),]$mean.quot.log2

mcols(dnam_res_sites_annotated)$diffmeth.p.adj.fdr <-
  dnam_res_sites[(dnam_res_sites$diffmeth.p.adj.fdr < 0.05),]$diffmeth.p.adj.fdr

seqlevelsStyle(dnam_res_sites_annotated) <- "NCBI"
```


Selecting transposable elements
This paper is source for gtf file:
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4757950/ 
And the download  link:
https://labshare.cshl.edu/shares/mhammelllab/www-data/TEtranscripts/TE_GTF/GRCh37_Ensembl_rmsk_TE.gtf.gz 
```{r}
TE <- read_gff(file = file.path(project_dir, "data/input/TE/GRCh37_rmsk_TE.gtf"))


TE_sub <- subsetByOverlaps(TE, dnam_res_sites_annotated)
rm(TE)


dnam_sites_TE <- join_overlap_left(dnam_res_sites_annotated, TE_sub) %>% 
  unique()

rm(dnam_res_sites_annotated)
rm(TE_sub)
```

## Saving DM cpg as csv
```{r}
rownames(dnam_res_sites) <- rownames(rnb_set@meth.sites)

write.table(dnam_res_sites, quote= FALSE, file = file.path(project_dir, "data/output/tables/dnam_res_sites.csv"))
```



## TE Bar plots
Only TE. from the sites
```{r}
dnam_te <- as.data.frame(mcols(dnam_sites_TE)) %>%
  mutate(status = ifelse(mean.quot.log2 > 0, "hypermethylated", "hypomethylated"))

heat_colors <- colorRampPalette(brewer.pal(n=9, name="RdBu"))(7)

ggplot(dnam_te, aes(x=class_id, fill = status)) +
  geom_bar(position=position_dodge()) +
  scale_y_log10() +
  scale_fill_manual(values=c(heat_colors[c(1,6)]))+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+
  theme(#panel.grid = element_blank(),
         # panel.background = element_rect(fill = brewer.pal(3,"Greys")[1]),
     panel.background = element_rect(fill = "white"),
         panel.border = element_rect(fill = "transparent"),
         panel.grid = element_line(color = brewer.pal(5,"Greys")[1]),
         legend.key=element_blank(),
         axis.title=element_text(size=10),
         legend.text=element_text(size=10),
         legend.title=element_text(size=10))
  


```
Adding genes and promoter information 
```{r}
rnb_genes <- annotation(rnb_set, type = "genes") %>%
  dplyr::select(Chromosome, Start, End, symbol) %>%
  as_granges(seqnames = Chromosome, start= Start, end = End, keep_mcols = TRUE) %>%
  unique()

seqlevelsStyle(rnb_genes) <- "NCBI"

dnam_sites_TE <- join_overlap_left(dnam_sites_TE, rnb_genes) %>% unique()

rm(rnb_genes)

rnb_promoters <- annotation(rnb_set, type = "promoters") %>%
  dplyr::select(Chromosome, Start, End, promoter_symbol = "symbol") %>%
  as_granges(seqnames = Chromosome, start= Start, end = End, keep_mcols = TRUE) %>%
  unique()

seqlevelsStyle(rnb_promoters) <- "NCBI"

dnam_sites_TE <- join_overlap_left(dnam_sites_TE, rnb_promoters) %>% unique()

rm(rnb_promoters)

```


```{r}
mcols(dnam_sites_TE) %>%
  as.data.frame() %>%
  mutate(isTE = ifelse(is.na(class_id), 0,1 ),
         isGene = ifelse(is.na(symbol),0,1)) %>%
  select(isTE, isGene) %>%
  table()
```

```{r, eval=FALSE}

# When dplyr fails for steps in
for( i in 1:length(dnam_sites_TE)){
  if(!is.na(mcols(dnam_sites_TE)$class_id[i])){ 
    cpg_annot <- append(cpg_annot, mcols(dnam_sites_TE)$class_id[i])}
  
  else if (!is.na(mcols(dnam_sites_TE)$symbol[i])){
    cpg_annot <- append(cpg_annot, "gene_body")
  }
  else if (!is.na(mcols(dnam_sites_TE)$promoter_symbol[i])){
    cpg_annot <- append(cpg_annot, "promoter_region")
  }
  else{
    cpg_annot <- append(cpg_annot, NA)
  }

 if(i %% 10000 == 0)
  print(i)
}

saveRDS(cpg_annot, file= file.path(project_dir, "data/processed/cpg_annot_vec.rds"))
```


```{r}
cpg_annot <- readRDS(file= file.path(project_dir, "data/processed/cpg_annot_vec.rds"))

dnam_sites_TE$cpg_annotations <- cpg_annot
```


```{r}
dnam_te <- as.data.frame(mcols(dnam_sites_TE)) %>%
  mutate(status = ifelse(mean.quot.log2 > 0, "hypermethylated", "hypomethylated")) %>%
  # Selecting only annotations that I can interpret
  filter(is.element(cpg_annotations, c("DNA", "gene_body", "LINE", "LTR",
                                       # "Other","RC", "RNA",
                                        "promoter_region", "Satellite", "SINE"))) %>%
  dplyr::mutate(cpg_annotations = ifelse(cpg_annotations=="promoter_region", "Promoter region", 
                                         ifelse(cpg_annotations=="gene_body", "Gene body", cpg_annotations)))
  

chi_tbl <- dnam_te %>% dplyr::select(cpg_annotations, status) %>% table()
chi <- chisq.test(chi_tbl)
# OVo je za DM CGI-jeve na temelju tablice
chisq.test(x=c(5457, 6557))

dnam_te$cpg_annotations <- factor(dnam_te$cpg_annotations, levels = c("Gene body", "Promoter region", "LINE", "SINE", "LTR", "DNA", "Satellite")) 

scientific_10 <- function(x) {
  parse(text=gsub("e", " %*% 10^", scales::scientific_format()(x)))
}

ggplot(dnam_te, aes(x=cpg_annotations, fill = status)) +
  geom_bar(position=position_dodge(), color = "black", size = 0.5) +
  # scale_y_log10() +
  scale_y_continuous(trans = "log10", label=scientific_10) +
  scale_fill_manual(values=c(heat_colors[c(1,6)]))+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+
  theme(#panel.grid = element_blank(),
         # panel.background = element_rect(fill = brewer.pal(3,"Greys")[1]),
     panel.background = element_rect(fill = "white"),
         panel.border = element_rect(fill = "transparent"),
         panel.grid = element_line(color = brewer.pal(5,"Greys")[1]),
         legend.key=element_blank(),
         axis.title=element_text(size=10),
         legend.text=element_text(size=10),
         legend.title=element_text(size=10)) +
  labs(x = "", y = "", fill = "CpG status")

ggsave(width = 7, height = 5, filename = file.path(project_dir, "data/output/figures/DM_sites_TE.png"))
```

# DM CGIs
## Load data
```{r}
dnam_res_sites <- readRDS(file.path(project_dir,"data/processed/dnam_res_cgi.rds"))
```

## Saving DM CGI as csv
Preping for saving this as csv 
supplemental table
```{r}
names(dnam_res_sites) <- 1:length(dnam_res_sites)

data_ <- as.data.frame(dnam_res_sites) %>%
  mutate(CGI_coordinates = str_c(seqnames,":", start,"-", end)) %>%
  dplyr::select(ensembl, symbol, CGI_coordinates, log.quot, pvalue, padj, num.sites)


write.table(data_, quote= FALSE, row.names = FALSE, file = file.path(project_dir, "data/output/tables/dnam_res_cgi.csv"))
```



# DE mRNA in HPV
Treba ih poredati po abs FC
```{r}
mRNA_res_HPV <- read.csv(file = file.path(project_dir, "data/processed/mRNA_res_HPV.csv"))

data_ <- mRNA_res_HPV %>%
  arrange(desc(abs(log2FoldChange))) %>%
  dplyr::select(-X)
rownames(data_) <- c(1:length(data_$row))

data_1 <- data_[c(1:18),]

data_2 <- data_ %>%
  # filter(is.element(symbol, c("ZNF541", "PAX8", "TYMS", "ADCYAP1")))
  filter(is.element(symbol, c("ZNF541", "TYMS")))
data_ <- rbind(data_1, data_2)

data_ <- data_ %>%
    # arrange(desc(log2FoldChange))
  arrange(log2FoldChange)

data_$symbol <- factor(data_$symbol, levels = data_$symbol)




ggplot(data_, aes(#x = as.numeric(rownames(data_)),
                  x = log2FoldChange,
                  y= as.factor(symbol), 
                  size = -log(padj),
                  color = log2FoldChange)) +
  geom_point() +
  # scale_x_continuous(breaks = c(-5, -4, -3, -2,2, 4, 5)) +
  # scale_color_gradient(low = "red", high = "darkblue") +
  scale_color_gradientn(colours = brewer.pal(n=9, name="RdBu"))+
  theme_bw() +
  labs(y ="",
       x = "Log2 Fold Change",
       size = "-Log2 p-value",
       color = "Log2 Fold Change")+
  theme(axis.text.y = element_text(face="bold"))

ggsave(height = 5, width = 5, file.path(project_dir, "data/output/figures/mRNA_HPV_top20_genes_dotplot_v2.png"))

  
  
```


# miR in multiple clusters
I double checked and this matches the kmeans clusters as per: 20221110_1148__11clusters_100nstart_scaled 
```{r}
reg_df <- readxl::read_xlsx(file.path(project_dir, "data/output/tables", "integrated_table_per_GENE_clusters.xlsx"))
```


```{r}
message(paste(filter(reg_df, kmeans == 3 & !is.na(mir))$mir %>% unique() %>% length(), "-> number of mirs in cluster 3"))
message(paste(filter(reg_df, kmeans == 9 & !is.na(mir))$mir %>% unique() %>% length(), "-> number of mirs in cluster 9"))

# Overexpressed mirs
is.element(unique(filter(reg_df, kmeans == 3)$mir), filter(reg_df, kmeans == 9)$mir) %>% table()
message(paste(calculate_shared_percentage(unique(filter(reg_df, kmeans == 3)$mir),  filter(reg_df, kmeans == 9)$mir),"- percent of uniqe mirs that overlap"))
message(paste(filter(reg_df, kmeans == 5 & !is.na(mir))$mir %>% unique() %>% length(), "-> number of mirs in cluster 5"))
message(paste(filter(reg_df, kmeans == 7 & !is.na(mir))$mir %>% unique() %>% length(), "-> number of mirs in cluster 8"))
# Underexpressed mirs
is.element(unique(filter(reg_df, kmeans == 7)$mir), filter(reg_df, kmeans == 5)$mir) %>% table()
message(paste(calculate_shared_percentage(unique(filter(reg_df, kmeans == 7)$mir),  filter(reg_df, kmeans == 5)$mir),"- percent of uniqe mirs that overlap"))


filter(reg_df, kmeans == 3)$mir %>% table() %>% as.data.frame() %>% arrange(Freq)
filter(reg_df, kmeans == 9)$mir %>% table() %>% as.data.frame() %>% arrange(Freq)
```
### E2F gene expressions
```{r}
e2f_ids <- data.frame(symbol = c("E2F1", "E2F2", "E2F3", "E2F4", "E2F5", "E2F6", "E2F7", "E2F8"),
                      ensembl = c("ENSG00000101412", "ENSG00000007968", "ENSG00000112242", "ENSG00000205250", "ENSG00000133740", "ENSG00000169016", "ENSG00000165891", "ENSG00000129173"))

# Preparing count data
data_ <- counts(mRNA_dds, normalized = TRUE)[e2f_ids$ensembl,] %>%
  as.data.frame() %>%
  rownames_to_column(var = "genes")
# Pivoting to long format
data_ <- pivot_longer(data_, cols = -genes, names_to = "samples", values_to = "counts")

# grabing sample annotation from dds object
data_ <-  data_ %>%
    mutate(sample_type = colData(mRNA_dds)[data_$samples, ]$Hist) %>%
  # adding symbol names
  mutate(gene_symbols = e2f_ids[match(genes, e2f_ids$ensembl),"symbol"])

```
Ploting
```{r}

col_pca <- c(
  # Dark red
  brewer.pal(8, "Reds")[7],
  brewer.pal(8, "Blues")[7]
)

ggplot(data_, aes(x = gene_symbols, y = counts, fill = sample_type)) +
  geom_boxplot() + 
  theme_bw() +
  ylab("Normalized counts")+
  xlab("") +
  labs(fill = "Sample group") + 
  theme(axis.title=element_text(size=18),
         legend.text=element_text(size=18),
         legend.title=element_text(size=18),
        axis.text.x = element_text(size = 12))+
  scale_fill_manual(values = col_pca)

# ggsave(filename = file.path(project_dir, "data/output/figures/E2F_counts.png"),        height = 6, width = 9)

```
# Legend for sample plot
### Samples
```{r}
# Outcome <- c("A", "B", "C", "D",'E')
Data <- c("Available", "Not available")
shots <- rep('Label',2)
xc <-c(1:2)
yc <-c(1:2)
df <-data.frame(shots,Data,xc,yc)

palette_color <- c("A"='chartreuse3', "B"='gainsboro', "C"='dodgerblue3', "D"='firebrick2', "E"='cornsilk4')

ggplot(df, aes(x = xc, y = yc, fill = Data)) +
  geom_point(shape=22, alpha=0, size = 5) +
  # geom_blank() +
  scale_fill_manual(values=c("white", "gray"), drop=FALSE) +
  guides(fill = guide_legend(override.aes = list(alpha=1))) +
  theme_classic() +
  # theme(legend.position = c(0.5,0.5),
  #       legend.text = element_text(size = 10),
  #       legend.title = element_text(size = 10),
  #       legend.key = element_rect( color = NA, size = 5))
    theme(
    legend.position = c(0.5, 0.5),
    legend.text = element_text(size = 20),
    legend.title = element_text(size = 20),
    legend.key.size = unit(2, "lines")  # Adjust size here
  )

ggsave(height = 10, width = 10, filename = file.path(project_dir,"data/output/figures", "sample_table_legend_samples.pdf"))

```
### Location
```{r}
# Outcome <- c("A", "B", "C", "D",'E')
Location <- c("O", "OP")
shots <- rep('Label',2)
xc <-c(1:2)
yc <-c(1:2)
df <-data.frame(shots,Location = Location,xc,yc)

palette_color <- c("A"='chartreuse3', "B"='gainsboro', "C"='dodgerblue3', "D"='firebrick2', "E"='cornsilk4')

ggplot(df, aes(x = xc, y = yc, fill = Location)) +
  geom_point(shape=22, alpha=0, size = 5) +
  # geom_blank() +
  scale_fill_manual(values=brewer.pal(8, "Paired")[c(8,7)], drop=FALSE) +
  guides(fill = guide_legend(override.aes = list(alpha=1))) +
  theme_classic() +
  # theme(legend.position = c(0.5,0.5),
  #       legend.text = element_text(size = 10),
  #       legend.title = element_text(size = 10),
  #       legend.key = element_rect( color = NA, size = 5))
    theme(
    legend.position = c(0.5, 0.5),
    legend.text = element_text(size = 20),
    legend.title = element_text(size = 20),
    legend.key.size = unit(2, "lines")  # Adjust size here
  )

ggsave(height = 10, width = 10, filename = file.path(project_dir,"data/output/figures", "sample_table_legend_location.pdf"))

```

### HPV
```{r}
# Outcome <- c("A", "B", "C", "D",'E')
HPV <- c("N", "P")
shots <- rep('Label',2)
xc <-c(1:2)
yc <-c(1:2)
df <-data.frame(shots,HPV,xc,yc)

palette_color <- c("A"='chartreuse3', "B"='gainsboro', "C"='dodgerblue3', "D"='firebrick2', "E"='cornsilk4')

ggplot(df, aes(x = xc, y = yc, fill = HPV)) +
  geom_point(shape=22, alpha=0, size = 5) +
  # geom_blank() +
  scale_fill_manual(values=brewer.pal(8, "Paired")[c(4,3)], drop=FALSE) +
  guides(fill = guide_legend(override.aes = list(alpha=1))) +
  theme_classic() +
  # theme(legend.position = c(0.5,0.5),
  #       legend.text = element_text(size = 10),
  #       legend.title = element_text(size = 10),
  #       legend.key = element_rect( color = NA, size = 5))
    theme(
    legend.position = c(0.5, 0.5),
    legend.text = element_text(size = 20),
    legend.title = element_text(size = 20),
    legend.key.size = unit(2, "lines")  # Adjust size here
  )

ggsave(height = 10, width = 10, filename = file.path(project_dir,"data/output/figures", "sample_table_legend_HPV.pdf"))

```

# Figure 2 gprofiler
Just redid this in "Exploratory_analysis_and_...." script
```{r}
GSEA_SOURCES <- c("GO:BP","KEGG","REAC", "TF")
mRNA_DE <- dplyr::filter(as.data.frame(mRNA_res), padj < 0.05)

gost_ <- gost(rownames(filter(as.data.frame(mRNA_res), padj < 0.05)),
              gsea_)


```


# Session Info
```{r Session info}
sessionInfo()
```
